pkg/storage/ingest/writer_test.go

```diff
@@ -723,6 +760,101 @@ func TestWriter_WriteSync(t *testing.T) {
 	})
 }
 
+func TestWriter_WriteSync_HighConcurrencyOnKafkaClientBufferFull(t *testing.T) {
+	const (
+		topicName     = "test"
+		numPartitions = 1
+		partitionID   = 0
+		numWorkers    = 50
+		tenantID      = "user-1"
+		testDuration  = 3 * time.Second
+	)
+
+	var (
+		done    = make(chan struct{})
+		workers = sync.WaitGroup{}
+
+		writeSuccessCount = atomic.NewInt64(0)
+		writeFailureCount = atomic.NewInt64(0)
+	)
+
+	createRandomWriteRequest := func() *mimirpb.WriteRequest {
+		// It's important that each request has a different size to reproduce the deadlock.
+		metricName := strings.Repeat("x", rand.IntN(1000))
+
+		series := []mimirpb.PreallocTimeseries{mockPreallocTimeseries(metricName)}
+		return &mimirpb.WriteRequest{Timeseries: series, Metadata: nil, Source: mimirpb.API}
+	}
+
+	// If the test is successful (no WriteSync() request is in a deadlock state) then we expect the test
+	// to complete shortly after the estimated test duration.
+	ctx, cancel := context.WithTimeoutCause(context.Background(), 2*testDuration, errors.New("test did not compete within the expected time"))
+	t.Cleanup(cancel)
+
+	// Estimate the size of each record written in this test.
+	writeReqRecords, err := marshalWriteRequestToRecords(partitionID, tenantID, createRandomWriteRequest(), maxProducerRecordDataBytesLimit)
+	require.NoError(t, err)
+	require.Len(t, writeReqRecords, 1)
+	estimatedRecordSize := len(writeReqRecords[0].Value)
+	t.Logf("estimated record size: %d bytes", estimatedRecordSize)
+
+	cluster, clusterAddr := testkafka.CreateCluster(t, numPartitions, topicName)
+	cfg := createTestKafkaConfig(clusterAddr, topicName)
+	cfg.ProducerMaxBufferedBytes = 10000
+	cfg.WriteTimeout = testDuration * 10 // We want the Kafka client to block in case of any issue.
+
+	// Throttle a very short (random) time to increase chances of hitting race conditions.
+	cluster.ControlKey(int16(kmsg.Produce), func(_ kmsg.Request) (kmsg.Response, error, bool) {
+		time.Sleep(time.Duration(rand.Int64N(int64(time.Millisecond))))
+
+		return nil, nil, false
+	})
+
+	writer, _ := createTestWriter(t, cfg)
+
+	// Start N workers that will concurrently write to the same partition.
+	for i := 0; i < numWorkers; i++ {
+		runAsync(&workers, func() {
+			for {
+				select {
+				case <-done:
+					return
+
+				default:
+					if err := writer.WriteSync(ctx, partitionID, tenantID, createRandomWriteRequest()); err == nil {
+						writeSuccessCount.Inc()
+					} else {
+						assert.ErrorIs(t, err, kgo.ErrMaxBuffered)
+						writeFailureCount.Inc()
+
+						// Stop a worker as soon as a non-expected error occurred.
+						if !errors.Is(err, kgo.ErrMaxBuffered) {
+							return
+						}
+					}
+				}
+			}
+		})
+	}
+
+	// Keep it running for some time.
+	time.Sleep(testDuration)
+
+	// Signal workers to stop and wait until they've done.
```

Comment:
[Nit]
```suggestion
	// Signal workers to stop and wait until they're done.
```

Reply:
Done in https://github.com/grafana/mimir/pull/8738/commits/c74391199b1270bc2f9bdb5ae55c1ca79c18d30e